IsInRangeConverter

GeorgeLeithead opened the pull request:
### Description of Change ###
Implement the proposed `IsInRangeConverter`.

 ### Linked Issues ###
#65 

 ### PR Checklist ###
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [x] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/176 <!-- Replace XYZ with your docs PR #. The checkbox will be automatically checked once the docs PR has been merged -->


 ### Additional information ###
Tested on:
- [x] Android
- [x] Windows
- [x] iOS
- [x] MacCatalyst

GeorgeLeithead commented:
> The issue is not championed

My appologies, I jumped the gun.  Was ensuring that before my time becomes more restricted starting next week, any work I have done was made available. 
 Thankfully the other 5 items I haven't created a PR yet, as they too have not yet been championed!

Have been discussing and sharing this with @bijington and didn't check to see if the championed status had been updated.

brminnick commented:
Closing PR until Proposal has been Approved

brminnick commented:
Reopening after Proposal Approval

GeorgeLeithead commented:
@JoonghyunCho @bijington This is ready for review.  Any suggestions, improvements, additions would be very much appreciated.  Documentation is also needed, which I will work on!

GeorgeLeithead commented:
> @JoonghyunCho @bijington This is ready for review. Any suggestions, improvements, additions would be very much appreciated. Documentation is also needed, which I will work on!

Please note that I do not have access to any iOS or Mac Catalyst devices (or Tizen), as such, please make sure that any reviewer who does have access, to please update as appropriate the checkbox indicating that it works on that platform.

JoonghyunCho commented:
I will check  if it works ok on iOS and Mac devices later tonight :)

GeorgeLeithead commented:
Also, please be aware that I also need to create the appropriate test cases.

JoonghyunCho commented:
@GeorgeLeithead Ok, I will see that together. It seems you already have many cases though.

GeorgeLeithead commented:
> @GeorgeLeithead Ok, I will see that together. It seems you already have many cases though.

Those are 'samples', which are there to try and provide the end developer with guidance and common usage patterns.  I need to write actual unit tests and I also like to make sure that there is great code coverage too!

GeorgeLeithead commented on src/CommunityToolkit.Maui/Converters/IsInRangeConverter.shared.cs:
```diff
@@ -0,0 +1,73 @@
+	public override object DefaultConvertReturnValue { get; set; } = new();
+
+	/// <summary>Minimum value.</summary>
+	public IComparable? Min { get; set; }
```
I can do that.

JoonghyunCho commented:
Looks great on the both iOS and maccatalyst! Just for the TimePicker on Mac has an issue opening on MAUI code. 
<img src="https://user-images.githubusercontent.com/14328614/200188244-7b905acc-d7fc-45b1-81d4-25e129763a68.png" width="250" /> <img width="300" alt="Screen Shot 2022-11-07 at 3 35 56 AM" src="https://user-images.githubusercontent.com/14328614/200188803-97814af7-20ff-4f92-a49f-7eb375095348.png">

JoonghyunCho commented:
Any plans for updating the tests cases?

GeorgeLeithead commented:
> Any plans for updating the tests cases?

@JoonghyunCho Absolutely.  Just a bit snowed under at this time, but hope to get to them soon, along with the documentation.

GeorgeLeithead commented:
@JoonghyunCho I have started to add the unit tests.  Still have another 160+ to add to cover all **appropriate** `IComparable` types.

pictos commented on src/CommunityToolkit.Maui/Converters/IsInRangeConverter.shared.cs:
```diff
@@ -0,0 +1,73 @@
+			throw new InvalidOperationException($"{nameof(TrueObject)} and {nameof(FalseObject)} should either be both defined or both omitted.");
+		}
+
+		if (MinValue is not null && MinValue.GetType() != value.GetType())
```
can we cache the `value.GetType()`? It will not change during the execution of the method

pictos commented on src/CommunityToolkit.Maui/Converters/IsInRangeConverter.shared.cs:
```diff
@@ -0,0 +1,73 @@
+			throw new ArgumentOutOfRangeException(nameof(value), $" is expected to be of type {nameof(MaxValue)}");
+		}
+
+		bool shouldReturnObjectResult = TrueObject is not null && FalseObject is not null;
```
```suggestion
		var shouldReturnObjectResult = TrueObject is not null && FalseObject is not null;
```

pictos commented on samples/CommunityToolkit.Maui.Sample/ViewModels/Converters/IsInRangeConverterViewModel.cs:
```diff
@@ -0,0 +1,24 @@
+	{
+		get
+		{
+			bool success = char.TryParse(StringCharInRange, out char returnChar);
```
```suggestion
			var success = char.TryParse(StringCharInRange, out var returnChar);
```

pictos commented on src/CommunityToolkit.Maui.UnitTests/Converters/IsInRangeConverterTests.cs:
```diff
@@ -0,0 +1,284 @@
+			TrueObject = trueObject,
+		};
+
+		_ = Assert.Throws<ArgumentOutOfRangeException>(() => ((ICommunityToolkitValueConverter)isInRangeConverter).Convert(value, typeof(object), null, CultureInfo.CurrentCulture));
```
A friendly reminder that we don't use the `_` (discard) pattern. Can remove this and the others that you added?

pictos reviewed (commented):
Hey @GeorgeLeithead I've some comments, most of them around code style, is it possible to disable your analyzers when working on MCT?

GeorgeLeithead commented:
@pictos @JoonghyunCho I have added loads of Unit Tests and added the Documentation.  I think this is all now good to go!  Looking forward to more great PR reviews.

JoonghyunCho reviewed (dismissed):
LGTM

brminnick reviewed (approved):
Thanks George! 

Just FYI - I updated the sample app to use a single `Grid` for its layout. The initial implementation had `Grid`s nested inside of `HorizontalStackLayout`s nested inside of `VerticalStackLayout`s.

> don't attempt to reproduce the appearance of a specific layout by using combinations of other layouts, as this results in unnecessary layout calculations being performed. For example, don't attempt to reproduce a [Grid](https://learn.microsoft.com/en-us/dotnet/api/xamarin.forms.grid) layout by using a combination of [StackLayout](https://learn.microsoft.com/en-us/dotnet/api/xamarin.forms.stacklayout) instances. 

https://learn.microsoft.com/xamarin/xamarin-forms/deploy-test/performance#choose-the-correct-layout

Because of the large number of Rows in the `Grid`, I ported the `IsInRangeConverterPage` to C# Markup which allows the naming of rows + columns. This makes it easier to assign UI controls to their respective column + row and helps avoid confusion amongst which row is which in the `Grid`.

brminnick reviewed (approved):
Approving again after adding the Bindable Properties defined in #65

GeorgeLeithead commented:
> Approving again after adding the Bindable Properties defined in #65

@brminnick I used the existing converters as a guide for implementation, especially `CompareConverter.shared.cs` as it was close to what `IsInRangeConverter` was to provide.  If I KNEW that Bindabale Properties was required, or given time to implement myself, I would have very VERY much have preferred to do so myself!
However, I do also find it very strange that Bindable Properties have been added to this `Converter` but not to the other existing converters, if they are now 'required'.  Especially for from a consistent end user experience.

GeorgeLeithead commented:
> Approving again after adding the Bindable Properties defined in #65

@brminnick I used the existing converters as a guide for implementation, especially `CompareConverter.shared.cs` as it was close to what `IsInRangeConverter` was to provide.  If I KNEW that Bindabale Properties was required, or given time to implement myself, I would have very VERY much have preferred to do so myself!
However, I do also find it very strange that Bindable Properties 

> Thanks George!
> 
> Just FYI - I updated the sample app to use a single `Grid` for its layout. The initial implementation had `Grid`s nested inside of `HorizontalStackLayout`s nested inside of `VerticalStackLayout`s.
> 
> > don't attempt to reproduce the appearance of a specific layout by using combinations of other layouts, as this results in unnecessary layout calculations being performed. For example, don't attempt to reproduce a [Grid](https://learn.microsoft.com/en-us/dotnet/api/xamarin.forms.grid) layout by using a combination of [StackLayout](https://learn.microsoft.com/en-us/dotnet/api/xamarin.forms.stacklayout) instances.
> 
> https://learn.microsoft.com/xamarin/xamarin-forms/deploy-test/performance#choose-the-correct-layout
> 
> Because of the large number of Rows in the `Grid`, I ported the `IsInRangeConverterPage` to C# Markup which allows the naming of rows + columns. This makes it easier to assign UI controls to their respective column + row and helps avoid confusion amongst which row is which in the `Grid`.

@brminnick Originally it started off as 1 sample that I hoped to make multi-use.  However it didn't pan out that way.  I should have converted to be a grid, and would have been happy to do the work, if I was given a chance to do so from a review.

brminnick commented:
No worries at all, George! You always do fantastic work and we're honored to have you as a strong contributor ❤️ 

The BindableProperties were in the API Design for the approved Proposal, and I wanted to ensure we honored that spec. I agree it is a bit inconsistent with other Converters today, but I think it opens up an awesome opportunity to add BindableProperties to existing Converters!

And your sample was great too! Since I know many devs will copy/paste our samples, I just wanted to ensure we start them off on the right path using best practices. You'd be surprised how many times I've been blamed in the past because I wrote some quick + dirty code as an example, only to find another dev copy/pasted it and made it the foundation of their app 😳

GeorgeLeithead commented:
> No worries at all, George! You always do fantastic work and we're honored to have you as a strong contributor ❤️
> 
> The BindableProperties were in the API Design for the approved Proposal, and I wanted to ensure we honored that spec. I agree it is a bit inconsistent with other Converters today, but I think it opens up an awesome opportunity to add BindableProperties to existing Converters!
> 
> And your sample was great too! Since I know many devs will copy/paste our samples, I just wanted to ensure we start them off on the right path using best practices. You'd be surprised how many times I've been blamed in the past because I wrote some quick + dirty code as an example, only to find another dev copy/pasted it and made it the foundation of their app 😳

@brminnick Have you seen all the Unit Tests I put in? :-D 
For the Bindable Properties, could be raised as a PROPOSAL, and be an opportunity to bump up the samples, and an opportunity to update the documentation to provide more extensive doc examples.
